File: 2023/swift/AdventOfCode2023/Sources/Day24.swift
import SwiftZ3

struct Day24: AdventDay {
  var input: String

  struct Point3D {
    let x: Double, y: Double, z: Double

    static func from(_ input: String) -> Self {
      let coordinates = input.ints()
      return Point3D(
        x: Double(coordinates[0]), y: Double(coordinates[1]), z: Double(coordinates[2]))
    }
  }

  struct Intersection {
    let x: Double, y: Double
  }

  struct Hailstone {
    let position: Point3D
    let velocity: Point3D

    var slope: Double {
      velocity.y / velocity.x
    }

    func intersection(with other: Hailstone) -> Intersection? {
      if velocity.x == 0 || other.velocity.x == 0 || slope == other.slope {
        return Optional.none
      }

      let x =
        ((other.position.y - other.slope * other.position.x) - (position.y - slope * position.x))
        / (slope - other.slope)
      let y = slope * (x - position.x) + position.y

      let t1 = (x - position.x) / velocity.x
      let t2 = (x - other.position.x) / other.velocity.x

      if t1 < 0 || t2 < 0 {
        return Optional.none
      }

      return Intersection(x: x, y: y)
    }

    static func from(_ input: String) -> Self {
      let split = input.split(separator: "@", maxSplits: 1)
      return Self(position: Point3D.from("\(split[0])"), velocity: Point3D.from("\(split[1])"))
    }
  }

  func part1() -> Any {
    let hailstones = input.lines().map { Hailstone.from($0) }
    let range = 200000000000000.0...400000000000000.0

    return
      hailstones
      .combinations(ofCount: 2)
      .compactMap { hailstones in hailstones[0].intersection(with: hailstones[1]) }
      .filter { intersection in range.contains(intersection.x) && range.contains(intersection.y) }
      .count
  }

  // Thank you https://www.reddit.com/r/adventofcode/comments/18pnycy/comment/kepl4p8/
  func part2() -> Any {
    let hailstones = input.lines().map { Hailstone.from($0) }

    let ctx = Z3Context()
    let solver = ctx.makeSolver()

    let x: Z3Real = ctx.makeConstant(name: "x")
    let y: Z3Real = ctx.makeConstant(name: "y")
    let z: Z3Real = ctx.makeConstant(name: "z")
    let vx: Z3Real = ctx.makeConstant(name: "vx")
    let vy: Z3Real = ctx.makeConstant(name: "vy")
    let vz: Z3Real = ctx.makeConstant(name: "vz")

    for (idx, hailstone) in hailstones[0...2].enumerated() {
      let t: Z3Real = ctx.makeConstant(name: "t\(idx)")
      let h_px = ctx.makeIntToReal(ctx.makeInteger64(Int64(hailstone.position.x)))
      let h_py = ctx.makeIntToReal(ctx.makeInteger64(Int64(hailstone.position.y)))
      let h_pz = ctx.makeIntToReal(ctx.makeInteger64(Int64(hailstone.position.z)))
      let h_vx = ctx.makeIntToReal(ctx.makeInteger64(Int64(hailstone.velocity.x)))
      let h_vy = ctx.makeIntToReal(ctx.makeInteger64(Int64(hailstone.velocity.y)))
      let h_vz = ctx.makeIntToReal(ctx.makeInteger64(Int64(hailstone.velocity.z)))

      solver.assert((x + vx * t) == (h_px + h_vx * t))
      solver.assert((y + vy * t) == (h_py + h_vy * t))
      solver.assert((z + vz * t) == (h_pz + h_vz * t))
    }

    if solver.check() != .satisfiable {
      fatalError("Failed to solve equation")
    }

    return solver.getModel()!.eval(x + (y + z))!.numeralDouble
  }
}

File: 2023/kotlin/src/Day24.kt
import com.microsoft.z3.Context
import com.microsoft.z3.Status

private operator fun <T> List<T>.component6() = this[5]

private data class Intersection(val x: Double, val y: Double)

private data class Hailstone(val position: Point3D<Long>, val velocity: Point3D<Long>) {
    private val slope = velocity.y / velocity.x.toDouble()

    fun intersection(other: Hailstone): Intersection? {
        if (velocity.x == 0L || other.velocity.x == 0L || slope == other.slope) return null

        val x = ((other.position.y - other.slope * other.position.x) - (position.y - slope * position.x)) / (slope - other.slope)
        val y = slope * (x - position.x) + position.y

        val t1 = (x - position.x) / velocity.x
        val t2 = (x - other.position.x) / other.velocity.x

        if (t1 < 0 || t2 < 0) return null

        return Intersection(x, y)
    }

    companion object {
        fun of(input: String) =
            input
                .split("@")
                .let { (position, velocity) -> Hailstone(Point3D.ofLong(position), Point3D.ofLong(velocity))
        }
    }
}

private fun part1(input: String): Int {
    val hailstones = input.lines().map { Hailstone.of(it) }
    val range = 200000000000000.0..400000000000000.0

    return hailstones
        .combinations(2)
        .mapNotNull { (h1, h2) -> h1.intersection(h2) }
        .count { (x, y) -> x in range && y in range }
}

// Thank you https://www.reddit.com/r/adventofcode/comments/18pnycy/comment/kepl4p8/
//
// MacOS Z3 setup, downloaded https://github.com/Z3Prover/z3/releases/tag/z3-4.13.0 release
// and extracted bin/{libz3.dylib,libz3java.dylib} to project root
private fun part2(input: String): Long {
    val hailstones = input.lines().map { Hailstone.of(it) }

    val ctx = Context()
    val solver = ctx.mkSolver()
    val (x, y, z, vx, vy, vz) = listOf("x", "y", "z", "vx", "vy", "vz").map { ctx.mkRealConst(it) }

    hailstones.take(3).forEachIndexed() { idx, hailstone ->
        /**
         * t >= 0
         * x + vx * t == hailstone.position.x + hailstone.velocity.x * t
         * y + vy * t == hailstone.position.y + hailstone.velocity.y * t
         * z + vz * t == hailstone.position.z + hailstone.velocity.z * t
         */
        val t = ctx.mkRealConst("t$idx")
        solver.add(
            ctx.mkEq(ctx.mkAdd(x, ctx.mkMul(vx, t)), ctx.mkAdd(ctx.mkReal(hailstone.position.x), ctx.mkMul(ctx.mkReal(hailstone.velocity.x), t))),
            ctx.mkEq(ctx.mkAdd(y, ctx.mkMul(vy, t)), ctx.mkAdd(ctx.mkReal(hailstone.position.y), ctx.mkMul(ctx.mkReal(hailstone.velocity.y), t))),
            ctx.mkEq(ctx.mkAdd(z, ctx.mkMul(vz, t)), ctx.mkAdd(ctx.mkReal(hailstone.position.z), ctx.mkMul(ctx.mkReal(hailstone.velocity.z), t)))
        )
    }

    if (solver.check() != Status.SATISFIABLE) {
        throw RuntimeException()
    }

    return solver.model.eval(ctx.mkAdd(x, ctx.mkAdd(y, z)), false).toString().toLong()
}

fun main() {
    val input = readInput("Day24")
    part1(input).println()
    part2(input).println()
}
